### PR TITLE
KAN Surface Decoder: learnable spline activations in output head

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The SurfaceRefinementHead currently uses standard MLP layers (Linear → ReLU → Linear) to map hidden features to surface pressure predictions. Kolmogorov-Arnold Networks (KANs) replace fixed activations with learnable B-spline functions on edges, enabling more expressive function approximation with the same parameter count.

**Key reference:** Liu et al., \"KAN: Kolmogorov-Arnold Networks\" (ICLR 2025, arXiv:2404.19756). Demonstrated 2-5x better accuracy per parameter on PDE-related tasks (Poisson, 2D flow) vs standard MLPs. The key advantage: spline activations can represent sharp, non-smooth function features (like stagnation-point pressure discontinuities) more compactly than piecewise-linear activations.

**Why the output head?** We've learned that backbone/attention modifications consistently hurt p_tan. Replacing only the output projection (SurfaceRefinementHead) is architecturally safe — it doesn't touch the Physics_Attention_Irregular_Mesh or K/V computation. The hidden representation stays identical; only the final mapping from hidden_dim → 3 channels changes.

**Hypothesis:** The model's surface pressure errors are concentrated at sharp physical features (stagnation point, suction peak, pressure recovery kink). A KAN layer in the output head can represent these sharp mappings more precisely than a ReLU MLP, since learnable splines adapt to the function shape during training.

## Instructions

Implement a lightweight KAN layer and use it in the SurfaceRefinementHead. Keep it minimal — do NOT change the Transolver backbone.

### Step 1: Add KANLayer class to train.py

```python
class KANLayer(nn.Module):
    """B-spline Kolmogorov-Arnold Network layer with residual linear.
    
    Replaces fixed activations with learnable spline functions on edges.
    Reference: Liu et al., KAN (ICLR 2025, arXiv:2404.19756)
    """
    def __init__(self, in_features, out_features, grid_size=5, spline_order=3):
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.grid_size = grid_size
        self.spline_order = spline_order
        
        # Grid of knots (fixed, not learned)
        grid = torch.linspace(-2, 2, grid_size + 2 * spline_order + 1)
        self.register_buffer('grid', grid)
        
        # B-spline coefficients (learned)
        self.coeff = nn.Parameter(
            torch.zeros(out_features, in_features, grid_size + spline_order)
        )
        # Residual linear (for stability — same as in original KAN paper)
        self.residual = nn.Linear(in_features, out_features)
        
        # Initialize: small random spline coeff, identity-like residual
        nn.init.kaiming_normal_(self.coeff, mode='fan_in')
        nn.init.kaiming_normal_(self.residual.weight, mode='fan_in')
        
    def b_splines(self, x):
        """Compute B-spline basis functions. x: [*, in_features]"""
        x = x.unsqueeze(-1)  # [*, in, 1]
        grid = self.grid  # [grid_size + 2*order + 1]
        
        # Expand x to grid dimensions
        # Use recursive B-spline definition (de Boor)
        bases = ((x >= grid[:-1]) & (x < grid[1:])).float()
        for k in range(1, self.spline_order + 1):
            left = (x - grid[:-(k+1)]) / (grid[k:-1] - grid[:-(k+1)] + 1e-8)
            right = (grid[k+1:] - x) / (grid[k+1:] - grid[1:-k] + 1e-8)
            bases = left * bases[..., :-1] + right * bases[..., 1:]
        return bases  # [*, in, grid_size + order]
    
    def forward(self, x):
        """x: [*, in_features] -> [*, out_features]"""
        orig_shape = x.shape
        x_flat = x.reshape(-1, self.in_features)  # [N, in]
        
        # B-spline term
        spline_bases = self.b_splines(x_flat)  # [N, in, G+K]
        # spline_out[n, out] = sum_in sum_g coeff[out, in, g] * bases[n, in, g]
        spline_out = torch.einsum('nig,oig->no', spline_bases, self.coeff)  # [N, out]
        
        # Residual linear term
        residual_out = self.residual(x_flat)  # [N, out]
        
        out = spline_out + residual_out
        return out.reshape(*orig_shape[:-1], self.out_features)
```

### Step 2: Add --kan_decoder flag

Add to argparse:
```python
parser.add_argument('--kan_decoder', action='store_true',
                    help='Replace SurfaceRefinementHead final linear with KAN layer')
parser.add_argument('--kan_grid_size', type=int, default=5,
                    help='KAN B-spline grid size (default 5)')
parser.add_argument('--kan_spline_order', type=int, default=3,
                    help='KAN B-spline polynomial order (default 3)')
```

### Step 3: Modify SurfaceRefinementHead to use KAN

In the SurfaceRefinementHead `__init__` and `forward`:

```python
class SurfaceRefinementHead(nn.Module):
    def __init__(self, hidden_dim, out_dim, n_layers=3, kan_decoder=False, 
                 kan_grid_size=5, kan_spline_order=3):
        super().__init__()
        self.kan_decoder = kan_decoder
        layers = []
        for i in range(n_layers - 1):
            layers.extend([nn.Linear(hidden_dim, hidden_dim), nn.ReLU()])
        # Final output layer: KAN or standard linear
        if kan_decoder:
            layers.append(KANLayer(hidden_dim, out_dim, 
                                   grid_size=kan_grid_size, 
                                   spline_order=kan_spline_order))
        else:
            layers.append(nn.Linear(hidden_dim, out_dim))
        self.net = nn.Sequential(*layers)
    
    def forward(self, x):
        return self.net(x)
```

Then when instantiating SurfaceRefinementHead, pass:
```python
srfhead = SurfaceRefinementHead(
    hidden_dim=args.surface_refine_hidden,
    out_dim=3,
    n_layers=args.surface_refine_layers,
    kan_decoder=args.kan_decoder,
    kan_grid_size=args.kan_grid_size,
    kan_spline_order=args.kan_spline_order,
)
```

**Note:** The aft_foil_srf flag creates a second SurfaceRefinementHead for the aft foil. Apply KAN to BOTH heads for consistency.

### Step 4: Run 2 seeds

```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --wandb_name "tanjiro/kan-decoder-s42" --seed 42 \
  --wandb_group kan-surface-decoder \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --kan_decoder --kan_grid_size 5 --kan_spline_order 3

# Repeat for seed 73: --seed 73 --wandb_name "tanjiro/kan-decoder-s73"
```

**Debug tip:** First run a 5-epoch debug run (add `--debug_epochs 5` or reduce via env var) to verify the KAN layer doesn't crash. Check that parameter count increases modestly (KAN adds ~out × in × (G+K) extra params to the final layer).

**If training diverges or crashes:** The B-spline basis can produce NaN if x is far outside the grid [-2, 2]. Add input clamping before the spline: `x = x.clamp(-2, 2)` before `b_splines(x)`.

## Baseline

Current best metrics (PR #2290, 2-seed avg):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37

**Reproduce baseline (without KAN):**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --wandb_name "tanjiro/baseline-check" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### KAN Surface Decoder — B-spline output layer in SurfaceRefinementHead

Replaced the final `nn.Linear` in both `SurfaceRefinementHead` and `AftFoilRefinementHead` with a `KANLayer` (learnable B-spline activations on edges, grid_size=5, spline_order=3, with residual linear path). Zero-initialized to match existing identity-at-start pattern.

### Metrics (2-seed average)

| Metric | Baseline (#2290) | KAN Decoder | Δ |
|--------|------------------|-------------|---|
| **p_in** | **11.742** | 12.700 | **+8.2%** ❌ |
| **p_tan** | **27.874** | 29.100 | **+4.4%** ❌ |
| **p_oodc** | **7.643** | 8.400 | **+9.9%** ❌ |
| **p_re** | **6.419** | 7.050 | **+9.8%** ❌ |
| val/loss | ~0.37 | 0.392 | +5.9% ❌ |

### Per-seed breakdown

| Metric | Seed 42 (4er1cvls) | Seed 73 (0qc824o5) |
|--------|-------------------|-------------------|
| p_in | 12.8 | 12.6 |
| p_tan | 28.9 | 29.3 |
| p_oodc | 8.3 | 8.5 |
| p_re | 6.9 | 7.2 |
| val/loss | 0.3925 | 0.3918 |
| Best epoch | 147 | 146 |
| Peak VRAM | 46.1 GB | 46.5 GB |

### Commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent tanjiro --wandb_name "tanjiro/kan-decoder-s42" --seed 42 \
  --wandb_group kan-surface-decoder \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --kan_decoder --kan_grid_size 5 --kan_spline_order 3

# Seed 73: same but --seed 73 --wandb_name "tanjiro/kan-decoder-s73"
```

### W&B runs
- Seed 42: `4er1cvls`
- Seed 73: `0qc824o5`

### What happened

**The KAN decoder regressed all metrics by 4-10%.** This is a clear negative result.

The hypothesis was that learnable B-spline activations could better represent sharp pressure features (stagnation points, suction peaks) compared to the standard linear output layer. In practice, the opposite occurred:

1. **The final linear layer is already well-suited**: The SurfaceRefinementHead uses 3 hidden layers of LayerNorm + GELU before the output projection. By the time features reach the final layer, they are already highly processed — the mapping from 192-dim hidden → 3 outputs is relatively simple. A linear layer handles this efficiently; adding spline complexity doesn't help and likely overfits or introduces optimization difficulties.

2. **Spline parameters add noise to a well-conditioned problem**: The KAN layer adds `out × in × (grid_size + spline_order) = 3 × 192 × 8 = 4,608` spline coefficients on top of the residual linear. For a 192→3 mapping, this is significant parametric overhead for a simple output projection. The extra parameters likely don't have enough gradient signal to learn useful representations, given that the hidden layers have already done the heavy lifting.

3. **The zero-init start is maintained** (verified via debug runs — correction magnitudes start near zero), but as training progresses the spline path may introduce instability in the optimization landscape compared to the clean linear gradient flow.

4. **VRAM was modestly higher** (46.1-46.5 GB vs baseline ~44 GB), but not a problem on 96 GB GPUs.

### Suggested follow-ups

- The output head is not the bottleneck for surface pressure accuracy. Future improvements should focus elsewhere (backbone representation quality, loss formulation, data augmentation).
- If KAN is to be revisited, it may be more effective as an intermediate layer replacing GELU activations in the hidden layers (not just the output), but the general pattern of backbone/attention modifications hurting p_tan suggests caution.
- The consistent ~5-10% regression across all splits suggests this is a systematic issue with the approach, not a hyperparameter tuning problem.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2325 because ADVISOR: Closing — KAN decoder regresses all metrics (p_in +8.2%, p_oodc +9.9%, p_tan +4.4%, p_re +9.8%).

The B-spline activations added complexity to the output head without benefit. The standard Linear → ReLU → Linear SRF head is already well-suited for the smooth mapping from hidden features to pressure — the sharp features in the pressure distribution are captured by the hidden representation, not the output projection.

**Lesson:** The output head is not the bottleneck. The model's errors come from the attention/representation stage, not from the final linear projection. Modifying only the output layer cannot fix representation-level issues.

Moving you to a fresh experiment..
